### ISSUE TITLE
time_t time(time_t*)

### ISSUE THREAD

**Issue by wintersteiger** [openenclave/openenclave#3516]
corelibc promises `time_t oe_time(time_t* tloc);` ([here](https://github.com/openenclave/openenclave/blob/03e07014e80d4894aee58d41216eeaa6d321a11d/include/openenclave/corelibc/time.h#L26)), but I can't find its implementation and there is no such symbol in the libraries. Is the implementation missing or should the prototype be removed from the headers?

**Comment by radhikaj** [comment]
This can be removed, oe_time in the same file can also be removed.  
OE_INLINE
time_t time(time_t* tloc)
{
    return oe_time(tloc);
}
We should look through history and see if this was ever implemented.

**Comment by anakrish** [comment]
Here is the callgraph for `time` function. It indicates that time function is being called by oecore (attestation) as well as be mbedtls. However, they seem to be using libc as opposed to corelibc (I'm not sure if that is recommended).

```
 time 0x6e150 /home/anakrish/work/fork2/3rdparty/musl/musl/src/time/time.c:4
  ├── oe_datetime_now 0x55e60 /home/anakrish/work/fork2/common/datetime.c:228
  │   └── oe_create_sgx_endorsements 0x5d2c0 /home/anakrish/work/fork2/common/sgx/endorsements.c:23
  │       └── oe_get_sgx_endorsements 0x5c0b0 /home/anakrish/work/fork2/common/sgx/endorsements.c:329
  │           ├── oe_verify_sgx_quote 0x19810 /home/anakrish/work/fork2/common/sgx/quote.c:428 ...
  │           ├── _verify_evidence 0x4b850 /home/anakrish/work/fork2/common/sgx/verifier.c:546 ...
  │           └── _get_evidence 0x4f610 /home/anakrish/work/fork2/enclave/sgx/attester.c:95 ...
  └── x509_get_current_time 0x91480 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509.c:904
      ├── mbedtls_x509_time_is_past 0x91370 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509.c:969
      │   ├── mbedtls_x509_crt_is_revoked 0x88420 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:1779
      │   │   └── x509_crt_verifycrl 0x8d350 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:1802 ...
      │   ├── x509_crt_verify_chain 0x89520 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2258
      │   │   └── mbedtls_x509_crt_verify_restartable 0x88730 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2532 ...
      │   ├── x509_crt_verifycrl 0x8d350 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:1802
      │   │   └── x509_crt_verify_chain 0x89520 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2258 ...
      │   └── x509_crt_find_parent_in 0x8dba0 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2004
      │       └── x509_crt_find_parent 0x8cd80 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2127 ...
      └── mbedtls_x509_time_is_future 0x91d70 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509.c:979
          ├── x509_crt_verify_chain 0x89520 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2258
          │   └── mbedtls_x509_crt_verify_restartable 0x88730 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2532 ...
          ├── x509_crt_verifycrl 0x8d350 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:1802
          │   └── x509_crt_verify_chain 0x89520 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2258 ...
          └── x509_crt_find_parent_in 0x8dba0 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2004
              └── x509_crt_find_parent 0x8cd80 /home/anakrish/work/fork2/3rdparty/mbedtls/mbedtls/library/x509_crt.c:2127 ...
```

**Comment by mingweishih** [comment]
Copy @CodeMonkeyLeet comments:

The presence of gmtime.c and the include/openenclave/corelibc/time.h (see #3516) both look like regressions. PR #3057 removed these functions as they were not used by oecore, and they were reintroduced in PR #3059, likely as the result of a bad merge.

**Comment by wintersteiger** [comment]
That's right, the mbedTLS and attestation code (endorsement extraction) calls this function. I hit this when I tried to make mbedTLS call the right function (via `mbedtls_platform_set_time`), because in SGX-LKL it ended up calling SGX-LKL's musl `time`. Since mbedTLS allows us to override it, we have only the one use in `datetime.c` to take care of, which we could replace with `oe_get_time` if we want to remove it.

**Comment by wintersteiger** [comment]
Closing; fixed by #3567.
